crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2330,34 +2473,84 @@ impl AuthorityPerEpochStore {
+            // executed from consensus task. At this point if the lock was not already provided
+            // above, we know we won't be transitioning state for this commit.
             self.get_reconfig_state_read_lock_guard().should_accept_tx()
         };
         let make_checkpoint = should_accept_tx || final_round;
         if make_checkpoint {
-            let pending_checkpoint = PendingCheckpoint {
+            // Filter out roots of any deferred tx.
+            for deferred in deferred_tx_roots {
+                roots.remove(&deferred);
+                randomness_roots.remove(&deferred);
+            }
+
+            // Generate pending checkpoint for regular user tx.
+            let commit_height = if self.randomness_state_enabled() {
+                commit_round * 2
+            } else {
+                commit_round
+            };
+            let pending_checkpoint = PendingCheckpointV2::V2(PendingCheckpointV2Contents {
                 roots: roots.into_iter().collect(),
                 details: PendingCheckpointInfo {
                     timestamp_ms: commit_timestamp,
-                    last_of_epoch: final_round,
-                    commit_height: commit_round,
+                    last_of_epoch: final_round && randomness_round.is_none(),
+                    commit_height,
                 },
-            };
+            });
 
             self.write_pending_checkpoint(&mut batch, &pending_checkpoint)?;
             checkpoint_service.notify_checkpoint(&pending_checkpoint)?;
+
+            // Generate pending checkpoint for user tx with randomness.
+            if let Some(randomness_round) = randomness_round {
+                randomness_roots.insert(TransactionKey::RandomnessRound(
+                    self.epoch(),
+                    randomness_round,
+                ));
+
+                let pending_checkpoint = PendingCheckpointV2::V2(PendingCheckpointV2Contents {
+                    roots: randomness_roots.into_iter().collect(),
+                    details: PendingCheckpointInfo {
+                        timestamp_ms: commit_timestamp,
+                        last_of_epoch: final_round,
+                        commit_height: commit_height + 1,
+                    },
+                });
+
+                self.write_pending_checkpoint(&mut batch, &pending_checkpoint)?;
+                checkpoint_service.notify_checkpoint(&pending_checkpoint)?;
+            }
         }
 
         batch.write()?;
 
+        // Once commit processing is recorded, kick off randomness generation.
+        if let Some(randomness_round) = randomness_round {
+            let epoch = self.epoch();
+            let randomness_manager = self
+                .randomness_manager
+                .get()
+                .expect("randomness manager should exist if randomness round is provided")
+                .clone();
+            // Spawn a task for this, because we can't await while holding `lock`.
```

Comment:
the task should no longer be needed since there is no await

Reply:
good catch, removed